lib/utils.js

```diff
@@ -1,252 +1,186 @@
-      superclass.defaults({
-        from: getDefaultAccount(), // Use web3.eth.defaultAccount as the from account for all transactions
+      contract.setProvider(myWeb3.currentProvider);
+      contract.defaults({
+        from: Utils.getDefaultAccount(),
         gas: config.get("gasLimit")
       });
+      return contract;
+    } catch (ex) {
+      return undefined;
+    }
+  }
 
-      this.contract = contract;
-      for (const i in this.contract) {
-        if (this[i] === undefined) {
-          this[i] = this.contract[i];
-        }
-      }
+  /**
+   * throws an exception when web3 cannot be initialized or there is no default client
+   */
+  static getWeb3() {
+    if (Utils._web3) {
+      return Utils._web3;
     }
 
-    static async new() {
-      return superclass.new().then(
-        contract => {
-          return new this(contract);
-        },
-        ex => {
-          throw ex;
-        }
+    let preWeb3;
+
+    if (typeof web3 !== "undefined") {
+      // Look for injected web3 e.g. by truffle in migrations, or MetaMask in the browser window
+      // Instead of using the injected Web3.js directly best practice is to use the version of web3.js we have bundled
+      // see https://github.com/MetaMask/faq/blob/master/DEVELOPERS.md#partly_sunny-web3---ethereum-browser-environment-check
+      preWeb3 = new Web3(web3.currentProvider);
+    } else if (Utils._alreadyTriedAndFailed) {
+      // then avoid time-consuming and futile retry
+      throw new Error("already tried and failed");
+    } else {
+      // No web3 is injected, look for a provider at providerUrl (which defaults to localhost)
+      // This happens when running tests, or in a browser that is not running MetaMask
+      preWeb3 = new Web3(
+        new Web3.providers.HttpProvider(config.get("providerUrl"))
       );
     }
 
-    static async at(address) {
-      return superclass.at(address).then(
-        contract => {
-          return new this(contract);
-        },
-        ex => {
-          throw ex;
-        }
-      );
+    if (!preWeb3) {
+      Utils._alreadyTriedAndFailed = true;
+      throw new Error("web3 not found");
     }
 
-    static async deployed() {
-      return superclass.deployed().then(
-        contract => {
-          return new this(contract);
-        },
-        ex => {
-          throw ex;
-        }
-      );
+    if (typeof window !== "undefined") {
+      // Add to window for easy use in the console
+      window.web3 = preWeb3;
     }
 
+    return (Utils._web3 = preWeb3);
+  }
+
+  /**
+   * @param tx The transaction
+   * @param argName The name of the property whose value we wish to return, from  the args object: tx.logs[index].args[argName]
+   * @param eventName Overrides index, identifies which log, where tx.logs[n].event  === eventName
+   * @param index Identifies which log, when eventName is not given
+   */
+  static getValueFromLogs(tx, arg, eventName, index = 0) {
```

Comment:
I think we could move this to the transaction return type class

Reply:
also. why call it ReturnType? why not just call it a Transaction?

Reply:
> why call it ReturnType? why not just call it a Transaction?

Because what it is is a _type_ (a class), not a transaction. Anyways, `Transaction` is already defined elsewhere, like ethereum-tx I think.

Further, its subclasses have other properties besides the tx.

I didn't like the "ReturnType" either.  I've changed it to "Result".  See what you think.

> I think we could move this to the transaction return type class

I am leaving `getValueFromLogs` in Utils while using it in the method that I added (at your good suggestion) to the base return type class.

Reply:
could be TransactionWrapper. or ExtendTransaction to match what we do with TruffleContract. not a big deal though